pkgs/test/nixpkgs-check-by-name/src/check_result.rs

```diff
@@ -0,0 +1,74 @@
+use crate::nixpkgs_problem::NixpkgsProblem;
+use itertools::concat;
+use itertools::{
+    Either,
+    Either::{Left, Right},
+    Itertools,
+};
+
+/// A type alias representing the result of a check, either:
+/// - Err(anyhow::Error): A fatal failure, typically I/O errors.
+///   Such failures are not caused by the files in Nixpkgs.
+///   This hints at a bug in the code or a problem with the deployment.
+/// - Ok(Left(Vec<NixpkgsProblem>)): A non-fatal problem with the Nixpkgs files.
+///   Further checks can be run even with this result type.
+///   Such problems can be fixed by changing the Nixpkgs files.
+/// - Ok(Right(A)): A successful (potentially intermediate) result with an arbitrary value.
+///   No fatal errors have occurred and no problems have been found with Nixpkgs.
+pub type CheckResult<A> = anyhow::Result<Either<Vec<NixpkgsProblem>, A>>;
```

Comment:
For others: I paired with @whentze today, where we concluded that this doesn't work out very well, in short, the `Result` type is made for early exits, but in our case `Err(NonFatal(_))` should _not_ exit early, which makes the code rather complicated.

Instead we ended up going for:
- A new type `Validation` (leaning onto Haskell's [`Validation`](https://hackage.haskell.org/package/validation-1.1.3/docs/Data-Validation.html), which inspired this in the first place) instead of `Either`
- Handling `Result`'s as they happen in the checks using `?`, so this complexity is not in the `Validation` functions

There are now somewhat ugly `Ok(...)`'s wrapping parts of the code, but it's not terrible imo :)